File: domain-driven-design/go-web/internal/infrastructure/repository/order_repository_impl.go
// 基础设施层（Infrastructure）：订单仓储实现
package repository

import (
	"fmt"
	"go-web-order/internal/domain/order"
)

// OrderRepositoryImpl 订单仓储实现，具体存储操作
type OrderRepositoryImpl struct {
  orders map[int]*order.Order // 存储订单的内存映射
}

// NewOrderRepositoryImpl 创建新的订单仓储实现
func NewOrderRepositoryImpl() *OrderRepositoryImpl {
  return &OrderRepositoryImpl{orders: make(map[int]*order.Order)}
}

// Save 保存订单
func (r *OrderRepositoryImpl) Save(order *order.Order) error {
  r.orders[order.ID] = order // 存储订单到内存中
  return nil
}

// FindByID 根据ID查找订单
func (r *OrderRepositoryImpl) FindByID(id int) (*order.Order, error) {
  order, exists := r.orders[id]
  if !exists {
    return nil, fmt.Errorf("订单 ID %d 不存在", id)
  }
  return order, nil
}

// FindAll 查找所有订单
func (r *OrderRepositoryImpl) FindAll() ([]*order.Order, error) {
  orders := []*order.Order{}
  for _, order := range r.orders {
    orders = append(orders, order)
  }
  return orders, nil
}

// Delete 删除订单
func (r *OrderRepositoryImpl) Delete(id int) error {
  if _, exists := r.orders[id]; !exists {
    return fmt.Errorf("订单 ID %d 不存在，无法删除", id)
  }
  delete(r.orders, id)
  return nil
}

// FindByCustomerName 根据客户名称查找订单
func (r *OrderRepositoryImpl) FindByCustomerName(customerName string) ([]*order.Order, error) {
  var result []*order.Order
  for _, order := range r.orders {
    if order.CustomerName == customerName {
      result = append(result, order)
    }
  }
  if len(result) == 0 {
    return nil, fmt.Errorf("没有找到客户名称为 %s 的订单", customerName)
  }
  return result, nil
}

File: domain-driven-design/java-web/src/main/java/com/javaweborder/infrastructure/repository/OrderRepositoryImpl.java
// 基础设施层（Infrastructure）：订单仓储实现
package com.javaweborder.infrastructure.repository;

import com.javaweborder.domain.order.Order;
import com.javaweborder.domain.order.OrderRepository;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

public class OrderRepositoryImpl implements OrderRepository {
    
    private final Map<Long, Order> orders = new ConcurrentHashMap<>(); // 线程安全的 HashMap

    @Override
    public void save(Order order) {
        if (orders.containsKey(order.getId())) {
            throw new NoSuchElementException("订单 ID " + order.getId() + " 已存在");
        }
        orders.put(order.getId(), order);
    }


    @Override
    public Optional<Order> findById(long id) {
        if (!orders.containsKey(id)) {
            throw new NoSuchElementException("订单 ID " + id + " 未找到");
        }
        return Optional.ofNullable(orders.get(id));
    }

    @Override
    public List<Order> findAll() {
        return new ArrayList<>(orders.values());
    }

    @Override
    public void delete(long id) {
        if (!orders.containsKey(id)) {
            throw new NoSuchElementException("订单 ID " + id + " 不存在，无法删除");
        }
        orders.remove(id);
    }

    @Override
    public List<Order> findByCustomerName(String customerName) {
        List<Order> result = new ArrayList<>();
        for (Order order : orders.values()) {
            if (order.getCustomerName().equals(customerName)) {
                result.add(order);
            }
        }
        if (result.isEmpty()) {
            throw new NoSuchElementException("没有找到客户名称为 " + customerName + " 的订单");
        }
        return result;
    }
}
